Support installation of recent versions of Nessus

jsf9k opened the pull request:
## 🗣 Description ##

This pull request:
* Adds automatic detection of the package format and platform architecture.
* Automatically determines the best Nessus package to install given the desired version, based on the Nessus package files that are uploaded to the S3 bucket.
* Adds support and testing for Amazon Linux 2 and Fedora distributions.
* Adds a `molecule` scenario to test installation of the latest version of Nessus.

## 💭 Motivation and context ##

Some assessment groups have indicated that they would like to have the most recent version of Nessus preinstalled in their AMIs.  Whatever is decided by the assessments tech council, this is a useful thing for this Ansible role to support.

## 🧪 Testing ##

All automated tests pass.

## ✅ Pre-approval checklist ##

- [x] This PR has an informative and human-readable title.
- [x] Changes are limited to a single goal - *eschew scope creep!*
- [x] *All* future TODOs are captured in issues, which are referenced in code comments.
- [x] All relevant type-of-change labels have been added.
- [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [x] All relevant repo and/or project documentation has been updated to reflect the changes in this PR.
- [x] Tests have been added and/or modified to cover the changes in this PR.
- [x] All new and existing tests pass.

jsf9k committed:
Add role variables for the architecture and Linux distribution

The default values match the previous behavior.

jsf9k committed:
Modify Nessus package pattern to match any distribution and architecture

jsf9k committed:
Add a scenario for testing installation of the latest version of Nessus

jsf9k committed:
Allow specification of the package type

e.g., rpm, deb, etc.

mcdonnnj commented on README.md:
```diff
@@ -64,9 +64,12 @@ None.
-| package_file | The name of the Nessus Debian package file. | `Nessus-{{ version }}-debian6_amd64.deb` | No |
-| version | The version number of the Nessus Debian package file. | `8.7.1` | No |
+| architecture | The processor architecture for which Nessus is to be installed. | `amd64` | No |
+| distribution | The Linux distribution on which Nessus is to be installed. | `amd64` | No |
```
```suggestion
| distribution | The Linux distribution on which Nessus is to be installed. | `debian6` | No |
```

mcdonnnj commented on defaults/main.yml:
```diff
@@ -1,9 +1,19 @@
 
-# The name of the Nessus Debian package file
-package_file: "Nessus-{{ version }}-debian6_amd64.deb"
+# The Linux distribution on which Nessus is to be installed
```
I think this should be rephrased to establish that it's a Tenable configuration thang. The 10.x deb files use `debian9` and 8.x use `debian6` even though they support the exact same Linux distributions as an example.

jsf9k replied:
Please see commit f3e829d.

mcdonnnj commented on defaults/main.yml:
```diff
@@ -1,9 +1,19 @@
-# The version number of the Nessus Debian package file
-version: "8.15.5"
+# The format/file extension of the desired Nessus package file
+# (e.g. deb, rpm)
```
I am reluctant to mention any files other than `deb` until there's testing and support in place for other platforms. There is a `RedHat.yml` vars file but it just points to `Debian.yml` and no testing is done for non-Debian-based platforms.

jsf9k replied:
This doesn't resolve this comment, but please see commit 07dff2e where I removed the  `RedHat.yml` file of mysterious provenance.

jsf9k replied:
- Please see commit 8065cd2, which removes the need for the user to set the `architecture` and `format` role variables, and thereby removes all mention of them from the README.
- Note also commit 17782be, which generalizes the way that the role was checking whether Nessus was already installed so that it will work for non-Debian distros as well.  I haven't actually added support for non-Debian platforms, but this commit makes that more of a possibility.
- The complication with adding support for non-Debian distributions is that Nessus itself uses different architecture names for different distributions (e.g., `amd64` vs. `x86_64`) _and_ different distribution names depending on the Nessus version (e.g., `debian6` vs. `debian9`).  What is needed is a dictionary that allows one to look up the desired architecture name depending on the `ansible_distribution` fact, and a more complicated dictionary structure (or a function, if I can figure out how to create such a thing in Terraform) to look up the desired distribution name depending on the desired Nessus version, the value of the `ansible_distribution`, and the desired architecture.  It seems simpler to avoid adding support for Fedora, etc. until it is actually needed, especially considering the upkeep required to these data structures every time a new version of Nessus is released.  Are you OK with moving on if I document this in an issue?

mcdonnnj replied:
Yes I was fine with removing any mention of support for non-Debian platforms until we explicitly supported it and just tracking that in an issue.

jsf9k replied:
Despite my reluctance to do so, I slowly but surely implemented support for our standard RedHat-based distros.  Can you do a re-review of this PR?

mcdonnnj reviewed (commented):
Excellent idea! I have some suggestions/thoughts for your consideration. I would also like an issue to ensure support, and add testing (default/latest), for each platform we would normally support aligned with the packages Tenable offers for Nessus: Amazon Linux 2, Debian (Debian/Kali), Fedora (RHEL is a separate package), and Ubuntu.

dav3r commented on molecule/latest/tests/test_latest.py:
```diff
@@ -0,0 +1,30 @@
+def test_nessus_version(host):
+    """Check that the correct version of Nessus was installed."""
+    cmd = host.run("/opt/nessus/sbin/nessusd --version")
+    assert " 10.3.0 " in cmd.stdout
+
+
+@pytest.mark.parametrize("f", ["/tmp/Nessus-10.3.0-debian9_amd64.deb"])
```
It's not a big deal since it's only in two places, but what about making a variable to hold the version number (`10.3.0`) so that we only have to change it in one place when they release a new version?  Is there a way to grab that version number from the Ansible variable in `molecule/latest/converge.yml`?

jsf9k replied:
Please see commit 0b62bee and let me know what you think.

I'd like to combine the two sets of very similar tests into a single file that is shared between the two `molecule` scenarios, but right now I don't see a clean way to do that.  Please let me know if you have any thoughts @dav3r and @mcdonnnj.

dav3r replied:
https://github.com/cisagov/ansible-role-nessus/commit/0b62bee1af9bab4ea456b9cc820e48bddd9d60f6 looks great to me.  I don't have any brilliant ideas to combine those two tests - maybe just create an issue for that and move on?

jsf9k replied:
Please see #22 and commit 6cb12af.

dav3r reviewed (approved):
Good stuff! 🚂 
I had one question/idea for your consideration.

jsf9k committed:
Remove RedHat vars file

We don't support any non-Debian distributions right now.  This file
is a mystery to me; it must be an historical artifact.

jsf9k committed:
Fix incorrect default value listed in documentation

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

jsf9k committed:
Create variables for the arch, distro, package format, and version

These will make it easier to update the tests going forward as new
versions are released.

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

jsf9k committed:
Clarify that the arch and distro are as specified in the filename of the Nessus package

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

jsf9k committed:
Automatically discover values for package format and architecture

jsf9k committed:
Generalize check for Nessus installation to non-Debian distros

jsf9k committed:
Add TODO comments referring to an issue for DRYing out the test code

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

jsf9k committed:
Add a variable to account for the different separation characters

For some reason the Nessus package filenames use periods to separate
the distribution from the architecture in the Nessus package filename
for RedHat-based distros, while they use underscores for the same
purpose for Debian-based distros.  This variable takes this into
account.

jsf9k committed:
Move some variables from defaults to vars

This allows us to avoid creating dictionaries with distro-specific
keys, and is a simpler way of achieving the same thing.

jsf9k committed:
Automatically determine the package filename

This can be done based on the packages that are uploaded to the 3rd
party S3 bucket.

jsf9k committed:
Escape periods in regexes

Also touch up the comment fro package_file_regex.

jsf9k committed:
Add support for RedHat distros (Amazon Linux and Fedora)

jsf9k committed:
Remove two role variables that have moved to vars

jsf9k committed:
Modify nessus_package_pattern to also match the RedHat package file names

dav3r reviewed (approved):
Nice work on this - I likes! 🥇

mcdonnnj commented on molecule/latest/converge.yml:
```diff
@@ -0,0 +1,10 @@
+      ansible.builtin.include_role:
+        name: ansible-role-nessus
+      vars:
+        distribution: "debian9"
```
I think this was missed when you were refactoring.

```suggestion
```

mcdonnnj commented on vars/Amazon.yml:
```diff
@@ -0,0 +1,23 @@
+# should be installed.  Note that we force the distribution name here,
+# because what Nessus uses ("amzn") does not match what Ansible uses
+# ("Amazon").
+package_file_regex: "Nessus-{{ version }}-amzn[0-9]*\\.{{ architectures[ansible_architecture] }}\\.{{ format }}"
```
In all of the `package_file_regex` definitions you use the `*` operator in the distribution specifier. Are you intentionally using "zero or more" as opposed to using `+` for "one or more"?

jsf9k replied:
Yes.  My thinking was that they may at some point have a single `Nessus-*-debian-*.deb` file for a particular version that supports all extant releases of that distribution.  Are you thinking that this could result in an error?

mcdonnnj replied:
No I just wanted to make sure it was intentional since all of the file patterns are `distribution\d+` for Linux installers right now.

mcdonnnj reviewed (approved):
Thanks for putting in the work to make this a comprehensive implementation 🏆 I have one small suggestion a question but otherwise LGTM ✔

jsf9k committed:
Remove distribution variable from latest scenario

This variable is no longer being used.

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>
